core/pta/attestation.c

```diff
@@ -0,0 +1,589 @@
+{
+	// TODO: Dummy function get_endorsement_key
+	// do something hardware specific
+	uint8_t end_key[] = {0x9a, 0x04, 0xaa, 0x18,
+			0x2d, 0x03, 0x96, 0x74,
+			0x70, 0x8c, 0xe8, 0x07,
+			0xed, 0x91, 0x4c, 0xd1,
+			0x53, 0xcd, 0x9d, 0xf7,
+			0x80, 0x5e, 0x61, 0x74,
+			0x2f, 0x0a, 0xe4, 0x12,
+			0x94, 0x75, 0x8d, 0xd3 };
+	memcpy(key, end_key, 32);
+	return TEE_SUCCESS;
+}
+
+static TEE_Result decrypt_priv_key(uint8_t *plain, size_t *plain_size)
+{
+	TEE_Result res = TEE_ERROR_GENERIC;
+	void *ctx = NULL;
+	size_t block_size = 0;
+	uint32_t algo = TEE_ALG_AES_CBC_NOPAD;
+	uint32_t mode = TEE_MODE_DECRYPT;
+	size_t aes_key_len = TEE_AES_MAX_KEY_SIZE;
+	uint8_t endorsement_key[aes_key_len];
+	uint8_t *dest_ptr = plain;
+	uint8_t *src_ptr = att_data.enc_priv;
+	uint32_t enc_size = att_data.enc_priv_size;
+
+	res = crypto_cipher_get_block_size(algo, &block_size);
+	if (res)
+		return res;
+
+	if (enc_size % block_size != 0)
+		return TEE_ERROR_GENERIC;
+
+	res = get_endorsement_key(endorsement_key);
+	if (res)
+		return res;
+
+	res = crypto_cipher_alloc_ctx(&ctx, algo);
+	if (res)
+		return res;
+
+	res = crypto_cipher_init(ctx, mode, endorsement_key,
+				 aes_key_len, NULL, 0,
+				 att_data.iv, att_data.iv_size);
+	if (res)
+		return res;
+
+	for (size_t i = 0; i < enc_size / block_size; i++) {
+		res = crypto_cipher_update(ctx, mode, false,
+					   src_ptr, block_size, dest_ptr);
+		if (res)
+			return res;
+		dest_ptr += block_size;
+		src_ptr += block_size;
+	}
+
+	crypto_cipher_final(ctx);
+	*plain_size = enc_size - plain[enc_size - 1];
+
+	return res;
+}
+
+static TEE_Result get_sha256(uint8_t *dst, uint8_t *src, size_t src_len)
+{
+	TEE_Result res = TEE_ERROR_GENERIC;
+	void *ctx = NULL;
+	size_t digest_len = TEE_SHA256_HASH_SIZE;
+
+	res = crypto_hash_alloc_ctx(&ctx, TEE_ALG_SHA256);
+	if (res)
+		return res;
+
+	res = crypto_hash_init(ctx);
+	if (res)
+		return res;
+
+	res = crypto_hash_update(ctx, src, src_len);
+	if (res)
+		return res;
+
+	res = crypto_hash_final(ctx, dst, digest_len);
+	if (res)
+		return res;
+
+	crypto_hash_free_ctx(ctx);
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result acipher_sign(const uint8_t *msg, const size_t msg_len,
+			       uint8_t *sig, size_t *sig_len)
+{
+	TEE_Result res = TEE_ERROR_GENERIC;
+	struct rsa_keypair key_pair;
+	uint8_t *d_ptr = NULL;
+	uint8_t *n_ptr = NULL;
+	uint8_t *e_ptr = NULL;
+	size_t d_size = 0;
+	size_t n_size = 0;
+	size_t e_size = 0;
+	size_t plain_priv_len = 0;
+	uint8_t *plain_priv = calloc(att_data.enc_priv_size, sizeof(uint8_t));
+
+	if (!plain_priv)
+		return TEE_ERROR_OUT_OF_MEMORY;
+
+	// decrypt the AIK
+	res = decrypt_priv_key(plain_priv, &plain_priv_len);
+	if (res != TEE_SUCCESS)
+		return res;
+
+	// use it to sign the msg
+	res = crypto_acipher_alloc_rsa_keypair(&key_pair, 0);
+	if (res)
+		return TEE_ERROR_SECURITY;
+
+	res = pub_asn1_decode(&n_ptr, &e_ptr, &n_size, &e_size);
+	if (res != TEE_SUCCESS)
+		return res;
```

Comment:
Memory leakage, `key_pair` isn't freed.

Reply:
Please correct me if  I am wrong, but there is no function like `crypto_acipher_free_rsa_keypair`. There is only one for the public key. So I have to free each bignum seperatly?
```suggestion
	crypto_bignum_free(key_pair->e);
	crypto_bignum_free(key_pair->d);
	crypto_bignum_free(key_pair->n);
	crypto_bignum_free(key_pair->p);
	crypto_bignum_free(key_pair->q);
	crypto_bignum_free(key_pair->qp);
	crypto_bignum_free(key_pair->dp);
```

Reply:
It's because that function hasn't been needed yet. Feel free to add one.

Reply:
Fix [here](https://github.com/OP-TEE/optee_os/pull/4011/commits/c3e8a401eb4aad9a07affde8d829d63979b9f53a), but the `crypto_acipher_free_rsa_keypair` problem, I will adress it tomorrow.

Reply:
`crypto_acipher_free_rsa_keypair`#4028